odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+            exactly zero for equal (but not identical) ``x`` and ``y``.
+
+            This option can only be used if ``exponent`` is 2.0.
+
+            Default: False.
+
+        kwargs :
+            Further keyword arguments are passed to the weighting
+            classes.
+
+        See Also
+        --------
+        rn : constructor for real tensor spaces
+        cn : constructor for complex tensor spaces
+        tensor_space :
+            constructor for tensor spaces of arbitrary scalar
+            data type
+
+
+        Notes
+        -----
+        - A distance function or metric on a space :math:`\mathcal{X}`
+          is a mapping
+          :math:`d:\mathcal{X} \\times \mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all space elements
+          :math:`x, y, z`:
+
+          * :math:`d(x, y) \geq 0`,
+          * :math:`d(x, y) = 0 \Leftrightarrow x = y`,
+          * :math:`d(x, y) = d(y, x)`,
+          * :math:`d(x, y) \\leq d(x, z) + d(z, y)`.
+
+        - A norm on a space :math:`\mathcal{X}` is a mapping
+          :math:`\| \cdot \|:\mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all
+          space elements :math:`x, y`: and scalars :math:`s`:
+
+          * :math:`\| x\| \geq 0`,
+          * :math:`\| x\| = 0 \Leftrightarrow x = 0`,
+          * :math:`\| sx\| = |s| \cdot \| x \|`,
+          * :math:`\| x+y\| \\leq \| x\| +
+            \| y\|`.
+
+        - An inner product on a space :math:`\mathcal{X}` over a field
+          :math:`\mathbb{F} = \mathbb{R}` or :math:`\mathbb{C}` is a
+          mapping
+          :math:`\\langle\cdot, \cdot\\rangle: \mathcal{X} \\times
+          \mathcal{X} \\to \mathbb{F}`
+          satisfying the following conditions for all
+          space elements :math:`x, y, z`: and scalars :math:`s`:
+
+          * :math:`\\langle x, y\\rangle =
+            \overline{\\langle y, x\\rangle}`,
+          * :math:`\\langle sx + y, z\\rangle = s \\langle x, z\\rangle +
+            \\langle y, z\\rangle`,
+          * :math:`\\langle x, x\\rangle = 0 \Leftrightarrow x = 0`.
+
+        Examples
+        --------
+        >>> space = NumpyTensorSpace((2, 3), dtype='float64')
+        >>> space
+        rn((2, 3))
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='complex64')
+        >>> space
+        cn((2, 3), 'complex64')
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='int64')
+        >>> space
+        tensor_space((2, 3), 'int')
+        """
+        TensorSpace.__init__(self, shape, dtype, order)
+
+        dist = kwargs.pop('dist', None)
+        norm = kwargs.pop('norm', None)
+        inner = kwargs.pop('inner', None)
+        weighting = kwargs.pop('weighting', None)
+        exponent = kwargs.pop('exponent', 2.0)
+        dist_using_inner = bool(kwargs.pop('dist_using_inner', False))
+
+        if (not self.is_numeric and
+                any(x is not None for x in (dist, norm, inner, weighting))):
```

Comment:
Well sure it is OK, but it is somewhat missleading regarding what it is, can't think of anything better atm though.